feat(aws-elasticache): add ServerlessCache, User, and UserGroup L2 Constructs

mazyu36 opened the pull request:
### Issue # (if applicable)

Closes #55.

### Reason for this change
To support Serveless Cache L2 Construct.

<!--What is the bug or use case behind this change?-->

### Description of changes
Ad the following Constructs
* ServerlessCache
* User
* UserGroup

<!--What code changes did you make? Have you made any important design decisions?-->

### Description of how you validated changes
Add unit tests and integ test.

<!--Have you added any unit tests and/or integration tests?-->

### Checklist

- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/open-constructs/aws-cdk-library/blob/main/CONTRIBUTING.md)

---
_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license_

mazyu36 commented on src/aws-elasticache/user-group.ts:
```diff
@@ -0,0 +1,162 @@
+    this.validateUserGroupId();
+
+    const userGroup = this.createResource(this, 'Resource', {
+      engine: Engine.REDIS,
```
This value only supports `redis`.
So I didn't expose this property.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticache-user.html#cfn-elasticache-user-engine

mazyu36 replied:
At the moment, a UserGroup with the `valkey` engine cannot be deployed, and the `valkey` value currently serves no purpose. 
How about adding a property in the future when the `valkey` value becomes meaningful?

* Deployment fails when the engine is set to `valkey`. I have submitted an issue regarding this (https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues/2199).
* According to the [AWS documentation](https://docs.aws.amazon.com/AmazonElastiCache/latest/dg/Clusters.RBAC.html), only `redis` is noted as being supported.

> When using RBAC with Valkey clusters, you will still need to assign users and user groups the engine “redis”.

go-to-k replied:
I see.

> How about adding a property in the future when the valkey value becomes meaningful?

Let's go with this plan.

go-to-k replied:
resolved for now.

mazyu36 commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,334 @@
+    this.validateAuthenticationSettings();
+
+    const user = this.createResource(this, 'Resource', {
+      engine: Engine.REDIS,
```
This value only supports `redis`.
So I didn't expose this property.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticache-user.html#cfn-elasticache-user-engine

go-to-k replied:
see above

mazyu36 commented on src/aws-elasticache/user-group.ts:
```diff
@@ -0,0 +1,162 @@
+    const userGroup = this.createResource(this, 'Resource', {
+      engine: Engine.REDIS,
+      userGroupId: this.physicalName,
+      userIds: Lazy.list({ produce: () => ['default', ...this.users.map(user => user.userId)] }),
```
I made it so that the `default` is automatically added because it must be included.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticache-usergroup.html#cfn-elasticache-usergroup-userids

tmokmss replied:
Just for confirmation, is this correct? I thought the doc means that you have to include a user id whose username is `default`, not that user id itself is `default`.

mazyu36 replied:
I apologize for my misunderstanding, but we need to include users with the user **name** `"default"` (not user id).
When deploying at L1 without username "default", I got the following error.

> ElastiCacheServerlessCacheStack | 21:09:59 | CREATE_FAILED        | AWS::ElastiCache::UserGroup           | UserGroup (UserGroupC755C9ED) Resource handler returned message: "User group needs to contain a user with the user name default. 

This is also described in the following docs:
https://docs.aws.amazon.com/AmazonElastiCache/latest/dg/Clusters.RBAC.html#Users-management

Therefore, it is not appropriate to automatically add users with user id "default". 
Instead, I think we need to implement validation to check whether a user with username "default" exists among the added users.

I'll fix it.

mazyu36 replied:
I've redesigned property referencing [Redshift's](https://github.com/open-constructs/aws-cdk-library/blob/main/src/aws-redshiftserverless/namespace.ts#L68) `defaultIamRole` and `iamRoles` property.

* Add `defaultUser' property and validation to check `defaultUser` name is `default`.
* Change `users` property to required and add validation to check `users` include `defaultUser`.

mazyu36 replied:
Thanks. 
I used the link as a reference to add validation and removed the defaultUser property.

mazyu36 commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,334 @@
+      engine: Engine.REDIS,
+      userId: this.physicalName,
+      userName: this.props.userName ?? this.physicalName,
+      accessString: this.props.accessString ?? 'off -@all',
```
This property is optional in CloudFormation, but if omitted, it results in an internal error during deployment. 
Therefore, the default value that is automatically set when creating from the management console is used.

`off -@all` disables all commands for the specified user. 

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticache-user.html#cfn-elasticache-user-accessstring

mazyu36 commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,334 @@
+      userId: this.physicalName,
+      userName: this.props.userName ?? this.physicalName,
+      accessString: this.props.accessString ?? 'off -@all',
+      authenticationMode: this.renderAuthenticationMode(),
```
[NoPasswordRequired](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticache-user.html#cfn-elasticache-user-nopasswordrequired) and [Passwords](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticache-user.html#cfn-elasticache-user-passwords) in CloudFormation will not be exposed because `passwords` and `no password required` can be configured through AuthenticationMode. 

In the L2 construct,[AuthenticationMode](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-elasticache-user-authenticationmode.html) is used to set either `IAM`, `passwords`, or `no password required`.

go-to-k replied:
Reopened this from resolved (see above) :)

go-to-k replied:
Confirmed, and will resolve

mazyu36 commented on src/aws-elasticache/serverless-cache.ts:
```diff
@@ -0,0 +1,509 @@
+    this.validateFinalSnapshotName();
+    this.validateUserGroup();
+
+    const serverlessCache = this.createResource(this, 'Resource', {
```
[CacheUsageLimits](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticache-serverlesscache.html#cfn-elasticache-serverlesscache-cacheusagelimits) is not exposed yet because the schema in the version of aws-cdk-lib currently used by the open constructs library differs from the latest version. It will be added once the latest version is adopted.

go-to-k replied:
Reopened this for myself from resolved (see above) :)

mazyu36 commented on src/aws-elasticache/daily-snapshot-time.ts:
```diff
@@ -0,0 +1,68 @@
+/**
```
`daily-snapshot-time.ts` is almost entirely reused from fsx's [maintenance-time.ts](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-fsx/lib/maintenance-time.ts).

mazyu36 commented on src/aws-elasticache/user-group.ts:
```diff
@@ -0,0 +1,166 @@
+   *
+   * \`userGroupId\` can have up to 40 characters.
+   *
+   * \`userGroupId\` must consist only of alphanumeric characters or hyphens, with the first character as a letter, and it can't end with a hyphen or contain two consecutive hyphens.
```
This constraint is quoted from the management console. It is not written in the [CFn documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticache-usergroup.html#cfn-elasticache-usergroup-usergroupid).

<img width="764" alt="image" src="https://github.com/user-attachments/assets/cdba97ed-9591-45de-bb00-94bbed7bfff6">

mazyu36 commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,335 @@
+   *
+   * \`userGroupId\` can have up to 40 characters.
+   *
+   * \`userId\` must consist only of alphanumeric characters or hyphens, with the first character as a letter, and it can't end with a hyphen or contain two consecutive hyphens.
```
This constraint is quoted from the management console. It is not written in the [CFn documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticache-user.html#cfn-elasticache-user-userid).

<img width="768" alt="image" src="https://github.com/user-attachments/assets/f22e8e4f-5244-4532-9e97-70db6576696b">

mazyu36 commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,335 @@
+  /**
+   * The username of the user.
+   *
+   * The name can have up to 120 characters, and must not contain spaces.
```
This constraint is quoted from the management console. It is not written in the [CFn documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticache-user.html#cfn-elasticache-user-username).

<img width="605" alt="image" src="https://github.com/user-attachments/assets/c2ce9c51-fa2d-4b8e-b45d-6339385b70ba">

mazyu36 commented on src/aws-elasticache/serverless-cache.ts:
```diff
@@ -0,0 +1,502 @@
+
+  /**
+   * The unique identifier of the serverless cache.
+   * The name can have up to 40 characters, and must not contain spaces.
```
This constraint is quoted from the management console. It is not written in the [CFn documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticache-serverlesscache.html#cfn-elasticache-serverlesscache-serverlesscachename).

<img width="599" alt="image" src="https://github.com/user-attachments/assets/7aab7481-0e5e-4847-af8f-501c8f8fb758">

mazyu36 commented on src/aws-elasticache/serverless-cache.ts:
```diff
@@ -0,0 +1,502 @@
+
+  /**
+   * A description of the serverless cache.
+   * The description can have up to 255 characters and must not contain < and > characters.
```
This constraint is quoted from the management console. It is not written in the [CFn documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticache-serverlesscache.html#cfn-elasticache-serverlesscache-description).

<img width="621" alt="image" src="https://github.com/user-attachments/assets/14292860-1aa5-4ecd-a791-af43e4a45ae1">

mazyu36 commented on src/aws-elasticache/serverless-cache.ts:
```diff
@@ -0,0 +1,502 @@
+  /**
+   * The version number of the engine the serverless cache is compatible with.
+   *
+   * @default MajorVersion.VER_7
```
I'm not sure this is true, because at the moment only VER_7 is available.

mazyu36 replied:
I see. I think it's better to make the majorVersion required.
I've updated.

go-to-k replied:
I'm okay. I'll leave it unresolved for now so that I can remember it later when I see this code again.

go-to-k replied:
Reopened this for myself from resolved (see above) :)

mazyu36 commented on src/aws-elasticache/serverless-cache.ts:
```diff
@@ -0,0 +1,502 @@
+   * The number of serverless cache snapshots the system will retain.
+   * To enable automatic backups, this property must be set.
+   *
+   * \`snapshotRetentionLimit\` must be between 1 and 35.
```
Value range is quoted from management console.

go-to-k replied:
Reopened this for myself from resolved :)

mazyu36 commented:
There are several validations and configuration values that are not documented in the CFn documentation but have been added based on observations from the management console. I've left these as review comments, so please close if there are no issues.

tmokmss commented on src/aws-elasticache/user-group.ts:
```diff
@@ -0,0 +1,167 @@
+      physicalName:
+        props.userGroupId ??
+        Lazy.string({
+          produce: () => Names.uniqueResourceName(this, { maxLength: 40, allowedSpecialCharacters: '-' }).toLowerCase(),
```
I prefer to add `separator :'-'` because the generated name will be more readable in lowercased letters. For example:
without separator: `stacknameconstructpathsrandomhash`
with separator:  `stackname-constructpaths-randomhash`

I know it is a matter of preference, but what do you think?

mazyu36 replied:
Thank you. I like your suggestion, so I will incorporate.

mazyu36 replied:
updated.

tmokmss commented on src/aws-elasticache/serverless-cache.ts:
```diff
@@ -0,0 +1,500 @@
+  /**
+   * The engine the serverless cache is compatible with.
+   */
+  readonly engine: Engine;
```
Shouldn't we default it to Valkey? :)

mazyu36 replied:
Personally, I would also prefer defaul "Valkey"... LoL.
However, I'm not sure if it's appropriate, so I'll wait for the maintainer's opinion.

go-to-k replied:
Reopened this for myself from resolved :)

tmokmss commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,336 @@
+  /**
+   * Access permissions string used for this user.
+   *
+   * @default - 'off -@all'
```
How about setting it to `on ~* +@all` by default? There can be tradeoffs between security and usability, but I would prefer that the default values just work.

mazyu36 replied:
I also like this suggestion, but I will wait for the maintainer's opinion.

go-to-k replied:
IMO, OCL requires secure constructs. So I may prefer more secure settings.

ref: https://www.open-constructs.org/

tmokmss reviewed (commented):
Leaving some comments as I touched Valkey Serverless recently. Thank you for the great work! @mazyu36

mazyu36 commented:
@tmokmss 
Thank you for the comments!
I've updated.

tmokmss commented:
@mazyu36 Thanks! I left some additional comment but LGTM overall. Kudos to you :)

go-to-k reviewed (changes requested):
I have read UserGroup and a part of User! I'll read the rest later.

go-to-k commented on src/aws-elasticache/user-group.ts:
```diff
@@ -0,0 +1,191 @@
+    });
+
+    this.userGroupArn = userGroup.attrArn;
+    this.userGroupId = userGroup.ref;
```
I haven't actually deployed it, so sorry if I'm wrong...!

mazyu36 replied:
I confirmed that I can get userGroupId with 'ref' in the code below. Please let me know if I have any misunderstanding.

```ts
const userGroup = new ocf.aws_elasticache.UserGroup(this, "UserGroup", {
  userGroupId: "my-user-group",
  users: [defaultUser, iamUser, noPasswordRequiredUser],
});

new cdk.CfnOutput(this, "UserGroupId", {
  value: userGroup.userGroupId,
});
```

<img width="511" alt="image" src="https://github.com/user-attachments/assets/aaf69cde-ea0a-4dba-ab37-acdf89596a09" />

go-to-k commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,364 @@
+   * @default - 'off -@all'
+   * @see https://docs.aws.amazon.com/AmazonElastiCache/latest/dg/Clusters.RBAC.html#Access-string
+   */
+  readonly accessString?: string;
```
I'm not familiar with the details of redis configuration, but is it possible to create something like an enum-like class to make this easier to configure (with custom configure)?

If this is a reckless request, feel free to let me know.

go-to-k replied:
Of course, I realize that it would be impossible to cover everything.

I'm thinking it would be good to have just a few common ways to specify (this could end up being just a wish...)

https://docs.aws.amazon.com/AmazonElastiCache/latest/dg/Clusters.RBAC.html#Access-string

mazyu36 replied:
Thank you. I'll look into what typical patterns might be possible.

mazyu36 replied:
After some consideration, I think it might be better to keep it as a string.

If we were to use an Enum, it would look something like the followings, granting permissions for operations on all keys. However, this might give the impression that wildcard keys are recommended.

```ts

export enum AccessString {
  /**
   * Full access to all keys and commands
   * - All key patterns (~*)
   * - All commands (+@all)
   */
  FULL_ACCESS = 'on ~* +@all',

  /**
   * Read-only access to all keys
   * - All key patterns (~*)
   * - Read commands only (+@read)
   */
  READ_ONLY = 'on ~* +@read',

  /**
   * Write-only access to all keys
   * - All key patterns (~*)
   * - Write commands only (+@write)
   */
  WRITE_ONLY = 'on ~* +@write',

  /**
   * Pub/Sub only access
   * - All key patterns (~*)
   * - Only Pub/Sub related commands (+@pubsub)
   */
  PUB_SUB = 'on ~* +@pubsub',

  /**
   * Disabled user state
   * - User is turned off
   * - No commands allowed (-@all)
   */
  DISABLED = 'off -@all',
}
```

I think there are many use cases where access restrictions are applied based on specific keys. 
Additionally, since this uses the Redis/Valkey notation, keeping it as a string might feel more familiar to users.

What do you think?

```sh
on ~cache:* +@read   # can only read keys with prefix `cache`
```

go-to-k replied:
> However, this might give the impression that wildcard keys are recommended.

I see. That's certainly true. Thanks for your thoughts. Let's keep this on string.

go-to-k replied:
Resolved for now.

mazyu36 commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,364 @@
+/**
+ * A new or imported User.
+ */
+export abstract class UserBase extends Resource implements IUser {
```
I think this should be private.
Thanks.

mazyu36 commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,364 @@
+   */
+  readonly accessString?: string;
+
+  /**
+   * Specifies the authentication type.
+   */
+  readonly authenticationType: AuthenticationType;
+
+  /**
+   * Passwords used for this user account.
+   * You can create up to two passwords for each user.
+   *
+   * You must set at least one password when `authenticatipnType` is set to `AuthenticationType.PASSWORD`
+   *
+   * @default - no passwords for this user
+   */
+  readonly passwords?: SecretValue[];
```
Thank you. I'll consider it.

mazyu36 replied:
I have refactored the code. 
Since the grant method is only valid for IAM users, I have implemented it only in the IamUser class.

go-to-k replied:
Thank you for the changes!

This is exactly what I wanted! I will gladly resolve this thread.

mazyu36 commented on src/aws-elasticache/user-group.ts:
```diff
@@ -0,0 +1,191 @@
+      return errors;
+    }
+
+    if (!userNamelist.includes('default')) {
+      errors.push('A user with the username `default` must be included in `users`.');
+    }
```
I tried it, and users with the same username could not be added to a user group.
I will consider adding this validation.

<img width="911" alt="image" src="https://github.com/user-attachments/assets/76af2e2e-866a-49d6-b71e-1c54a7791b6f" />

mazyu36 commented on src/aws-elasticache/user-group.ts:
```diff
@@ -0,0 +1,191 @@
+   * @param user the user to add
+   */
+  public addUser(user: IUser): void {
+    if (this.users.includes(user)) {
+      throw new Error(`An adding user is already included in the user group, ARN: ${user.userArn}.`);
+    }
```
Here are the results of my testing:

* Users with the same user ID cannot be created at all. I believe it is not possible to validate this.
* Users with the same username can be created, but it is not possible to add multiple users with the same username to a single user group. This is the same restriction I mentioned in my comment about the default user. I will consider adding this validation.

mazyu36 replied:
Updated a validation to check duplicate usernames.

mazyu36 commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,364 @@
+    });
+
+    this.userArn = user.attrArn;
+    this.userId = user.ref;
```
I confirmed that I can get userId with 'ref' in the code below. Please let me know if I have any misunderstanding.

```ts
const defaultUser = new ocf.aws_elasticache.User(this, "DefaultUser", {
  authenticationType: AuthenticationType.NO_PASSWORD_REQUIRED,
  userId: "new-default",
  userName: "default",
});

new cdk.CfnOutput(this, "UserId", {
  value: defaultUser.userId,
});
```

<img width="511" alt="image" src="https://github.com/user-attachments/assets/ae4bec02-a278-4225-824a-3c86169e50b6" />

go-to-k replied:
Thank you, I resolve it.

mazyu36 commented on src/aws-elasticache/user-group.ts:
```diff
@@ -0,0 +1,191 @@
+   *
+   * @attribute
+   */
+  readonly userGroupId: string;
```
Thanks. You are right.
I've removed the label.

mazyu36 commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,364 @@
+   *
+   * @attribute
+   */
+  readonly userName: string;
```
Thanks. I've removed the labels.

mazyu36 commented on src/aws-elasticache/user-group.ts:
```diff
@@ -0,0 +1,191 @@
+
+    if (!userNamelist || userNamelist.some(userName => Token.isUnresolved(userName))) {
+      return errors;
+    }
```
Thanks.
I've fixed.

go-to-k commented on src/aws-elasticache/user-group.ts:
```diff
@@ -0,0 +1,211 @@
+
+    const seenUsernames = new Set<string>();
+
+    for (const username of userNamelist) {
+      if (seenUsernames.has(username)) {
+        return [
+          `Duplicate username found in user group: \`${username}\` is duplicated. Each username must be unique within a user group.`,
+        ];
+      }
+      seenUsernames.add(username);
+    }
```
Since we are doing delayed evaluation with `addValidation`, how about returning all duplicate username combinations without immediately returning?

mazyu36 replied:
Make sense.
I modified the validation logic to return all duplicate usernames.

go-to-k commented on src/aws-elasticache/user-group.ts:
```diff
@@ -0,0 +1,211 @@
+  private validateDuplicateUsernames(): string[] {
+    const userNamelist = this.users.map(user => user.userName);
+
+    if (!userNamelist || userNamelist.some(username => Token.isUnresolved(username))) {
```
see: https://github.com/open-constructs/aws-cdk-library/pull/57/files#r1906876504

mazyu36 replied:
Removed `!userNamelist ` check.

go-to-k commented on src/aws-elasticache/user-group.ts:
```diff
@@ -0,0 +1,211 @@
+   * @param user the user to add
+   */
+  public addUser(user: IUser): void {
+    if (this.users.some(existingUser => existingUser.userName === user.userName)) {
+      throw new Error(`A user with username \`${user.userName}\` already exists in the user group.`);
+    }
```
This validation would not be necessary by creating `validateDuplicateUsernames` and adding the [check](https://github.com/open-constructs/aws-cdk-library/pull/57/files#r1909140704). I think it is more user-friendly to be able to check all the duplicate usernames, including users specified in props as well as users added by addUser.

go-to-k replied:
There is a way to validate it with the current code not `validateDuplicateUsernames ` if the adding users with the following code at line 113 in the constructor, but this cannot allow users to detect all the duplicates. Instead, it allows for early checks. Both have their good points, but I find it more user-friendly to be able to find all duplicates at once with delayed checking.

```ts
  constructor(scope: Construct, id: string, props: UserGroupProps) {
  // ...
    for (const user of this.props.users){
      this.addUser(user);
    }
```

mazyu36 replied:
I see.
I removed this validation.

mazyu36 commented on src/aws-elasticache/user-group.ts:
```diff
@@ -0,0 +1,191 @@
+
+    const errors: string[] = [];
+
+    if (!userNamelist || userNamelist.some(userName => Token.isUnresolved(userName))) {
```
Sorry, I've missed it.
I've fixed it.

go-to-k commented on src/aws-elasticache/serverless-cache.ts:
```diff
@@ -0,0 +1,557 @@
+/**
+ * A ElastiCache Serverless Cache
+ */
+export interface IServerlessCache extends IResource, aws_ec2.IConnectable {
```
Is the address and port of ReaderEndpoint not necessary?

Also, should `serverlessCacheName` be `@attribute`?

go-to-k commented on src/aws-elasticache/serverless-cache.ts:
```diff
@@ -0,0 +1,557 @@
+   */
+  metric(metricName: string, props?: MetricOptions): Metric;
+
+  /**
+   * Metric for the total number of bytes used by the data stored in your cache.
+   */
+  metricBytesUsedForCache(props?: MetricOptions): Metric;
+
+  /**
+   * Metric for the total number of ElastiCacheProcessingUnits (ECPUs) consumed by the requests executed on your cache
+   */
+  metricElastiCacheProcessingUnits(props?: MetricOptions): Metric;
```
It is difficult to add methods to interface later for backward compatibility. On top of that, is there any flavor to adding these to the interface? Do you want this to be particularly important compared to other metrics? If so, are there any other required metrics?

mazyu36 commented on src/aws-elasticache/serverless-cache.ts:
```diff
@@ -0,0 +1,534 @@
+    });
+
+    this.serverlessCacheArn = serverlessCache.attrArn;
+    this.serverlessCacheName = serverlessCache.ref;
```
I confirmed that the serverless cache name can be retrieved using `ref`.

go-to-k commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,404 @@
+ *     accessString: 'on ~* +@all',
+ *   },
+ * );
+ */
+export class IamUser extends UserBase {
```
```suggestion
 */

export class IamUser extends UserBase {
```

go-to-k commented:
(Sorry. The last comment with the new line was just a mistake. I deleted it, so don't worry about it.)

mazyu36 commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,404 @@
+/**
+ * Base properties for all user types
+ */
+export interface UserPropsBase {
```
> As far as the other examples in the aws-cdk-lib, would UserBaseProps be better?

I prefer `UserBaseProps`, so I fixed it.

> Also, is it good to do the following super process, this.props = props; and this.validateUserId(this.props.userId); in the Base class? (That would change the above discussion.)

I'm sorry, but I don't fully understand the comment.
Are you suggesting that the super call, along with this.props = props and this.validateUserId, should be consolidated in the base class to eliminate duplication across child classes?

> Perhaps some of the following processes (such as userName and authenticationMode) could also be brought to the parent class with the parent's methods, abstract methods and subtype methods like "Template Method Patterns" or "Factory Methods Patterns" in GoF design patterns?

I’m not entirely sure about this either.
Does the comment refer to the following?

* Moving all common processing to the base class to reduce duplication in the child classes.
* For example, making the rendering of the differing authenticationMode in the child classes an abstract method in the base class, to be implemented by the child classes.

> Should we expose the interface?

BaseProps is not needed to be exported. Removed `export`.

go-to-k replied:
> Are you suggesting that the super call, along with this.props = props and this.validateUserId, should be consolidated in the base class to eliminate duplication across child classes?

> Moving all common processing to the base class to reduce duplication in the child classes.

> For example, making the rendering of the differing authenticationMode in the child classes an abstract method in the base class, to be implemented by the child classes.

Yeah, that's what I said. But I actually don't look it with my editor, I'm sorry if there are wrong things in my comments.

mazyu36 replied:
The changes are quite extensive, and there are some parts that aren't working as expected, so I’m refactoring incrementally.

> Also, is it good to do the following super process, this.props = props; and this.validateUserId(this.props.userId); in the Base class? (That would change the above discussion.)

I’ve refactored this to consolidate the logic into the Base class. Note that I had to add the `export` to `BaseProps` since an error occurred when it was not exported.

```sh
[2025-01-11T11:52:45.757] [ERROR] jsii/compiler - Type model errors prevented the JSII assembly from being created
src/aws-elasticache/user.ts:105:29 - error JSII3001: Type "UserBaseProps" cannot be used as the property type because it is private or @internal

105   protected readonly props: UserBaseProps;
                                ~~~~~~~~~~~~~

  src/aws-elasticache/user.ts:53:1
     53 interface UserBaseProps {
        ~~~~~~~~~~~~~~~~~~~~~~~~~
     54   /**
        ~~~~~
    ... 
     67   readonly accessString?: string;
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     68 }
        ~
```

> Perhaps some of the following processes (such as userName and authenticationMode) could also be brought to the parent class with the parent's methods, abstract methods and subtype methods like "Template Method Patterns" or "Factory Methods Patterns" in GoF design patterns?

I’m working on this refactor, but some parts are not working as expected. Therefore, I plan to proceed with the additional refactoring after confirming that the changes above are correct.

mazyu36 replied:
> Perhaps some of the following processes (such as userName and authenticationMode) could also be brought to the parent class with the parent's methods, abstract methods and subtype methods like "Template Method Patterns" or "Factory Methods Patterns" in GoF design patterns?

I've refactored.

go-to-k replied:
Thank you for the perfect changes! Very good!

I resolve it.

go-to-k commented on src/aws-elasticache/serverless-cache.ts:
```diff
@@ -0,0 +1,534 @@
+}
+
+/**
+ * Represents a ElastiCache Serverless Cache construct in AWS CDK.
```
nit.

```suggestion
 * Represents an ElastiCache Serverless Cache construct in AWS CDK.
```

mazyu36 replied:
Updated including others as well

go-to-k commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,360 @@
+  readonly accessString?: string;
+}
+
+/**
+ * Base class for all user types
+ */
+abstract class UserBase extends Resource implements IUser {
```
The normal XxxBase class is a class for implementing abstract attributes and methods for imported resources. But now the class has `constructor`, `createResource`, etc., so it is different from the conventional base class.

ref: [Abstract Base](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md#abstract-base)

So I'm wondering if one of the following two would be better.

- Change the class name to `BaseUser` to match `BaseService` for `FargateService` in aws-ecs. This makes it clear that it is not a conventional base class.
- Create an abstract class `User` and move resource creation process (`constructor`, `createResource`, etc.), etc. to that class. This makes the responsibilities per class cleaner.
  - Needs to change the `UserBaseProps` to `UserProps`

I think the latter is better for maintainability and extensibility as software, but it would be a structure not often seen in CDK.

What do you think?

mazyu36 replied:
Thank you. I preferred the latter, so I revised it accordingly.

go-to-k commented on src/aws-elasticache/serverless-cache.ts:
```diff
@@ -0,0 +1,534 @@
+ * );
+ */
+export class ServerlessCache extends ServerlessCacheBase implements IServerlessCache {
+  /**
+   * The serverless cache ARN.
+   */
+  readonly serverlessCacheArn: string;
+  /**
+   * The serverless cache name.
+   */
+  readonly serverlessCacheName: string;
+  /**
+   * The DNS hostname of the cache node.
+   */
+  readonly endpointAddress: string;
+  /**
+   * The port number that the cache engine is listening on.
+   */
+  readonly endpointPort: number;
+
+  /**
+   * The connection object associated with the ElastiCache Serverless Cache.
+   */
+  public readonly connections: aws_ec2.Connections;
```
according to the `connections`:

```suggestion
  /**
   * The serverless cache ARN.
   */
  public readonly serverlessCacheArn: string;
  /**
   * The serverless cache name.
   */
  public readonly serverlessCacheName: string;
  /**
   * The DNS hostname of the cache node.
   */
  public readonly endpointAddress: string;
  /**
   * The port number that the cache engine is listening on.
   */
  public readonly endpointPort: number;

  /**
   * The connection object associated with the ElastiCache Serverless Cache.
   */
  public readonly connections: aws_ec2.Connections;
```

go-to-k commented on src/aws-elasticache/serverless-cache.ts:
```diff
@@ -0,0 +1,534 @@
+      throw new Error('`snapshotRetentionLimit` must be specified when `dailySnapshotTime` is set.');
+    }
+
+    if (snapshotRetentionLimit !== undefined) {
+      if (snapshotRetentionLimit < 1 || snapshotRetentionLimit > 35) {
+        throw new Error(`\`snapshotRetentionLimit\` must be between 1 and 35, got: ${snapshotRetentionLimit}.`);
+      }
+    }
```
If there are no other conditions, we should refrain from nesting `if` and keep it to one level. If there are more in the future, we will return then.

mazyu36 replied:
Updated if statement.

go-to-k commented on src/aws-elasticache/serverless-cache.ts:
```diff
@@ -0,0 +1,534 @@
+
+    if (Token.isUnresolved(finalSnapshotName)) return;
+
+    if (finalSnapshotName !== undefined) {
```
Can we return early if `finalSnapshotName === undefined`? (We can move it into `if (Token.isUnresolved(finalSnapshotName)) return;`.)

mazyu36 replied:
Updated if statement.

go-to-k commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,360 @@
+    };
+  }
+
+  /**
+   * Grant the given identity the specified actions
+   * @param grantee the identity to be granted the actions
+   * @param actions the data-access actions
+   *
+   * @see https://docs.aws.amazon.com/service-authorization/latest/reference/list_amazonelasticache.html
+   */
+  public grant(grantee: aws_iam.IGrantable, ...actions: string[]): aws_iam.Grant {
```
What do you think about creating an `IIamUser` like `IFargateService` so we can call the grant method on the imported resource?

```ts
export interface IIamUser extends IUser {
  // ...
  grant(grantee: aws_iam.IGrantable, ...actions: string[]): aws_iam.Grant;
```

In that case, we need to create a `fromXxx` method that returns the `IUser` in each subclass. In such cases, common processing seems to be handled by creating a common method in aws-ecs, for example. Attributes can also be changed for each class (instead, an attribute interface is also created for each class.)

- https://github.com/go-to-k/aws-cdk/blob/caf24150992f84e3b28eb8c02ee7dc2f5b5ca84f/packages/aws-cdk-lib/aws-ecs/lib/fargate/fargate-service.ts#L116
- https://github.com/go-to-k/aws-cdk/blob/caf24150992f84e3b28eb8c02ee7dc2f5b5ca84f/packages/aws-cdk-lib/aws-ecs/lib/base/from-service-attributes.ts#L31

But in that case, we must write the same code for the current `grant` method in `class Import` in `fromXxx` method. So I'm not sure this needs to go that far :)

(As a matter of fact, I came up with this idea when I noticed that both IamUser.fromAttributes and PasswordUser.fromAttributes return the same IUser.)

mazyu36 replied:
Thank you. I’ll consider the revision approach. Indeed, it seems more convenient if the imported IamUser can use the grant method.

mazyu36 replied:
I refactored the code referencing the ECS implementation. 
Please let me know if anything doesn't match your expectations.

* Prepared interfaces for each user type along with their attributes. IamUser can be imported with just UserId and also implements the grant method.
* Renamed the base user class from `User` to `BaseUser`
* Removed the `UserBase` class for importing as I think it's no longer necessary.

go-to-k replied:
Resolve this thread.

mazyu36 commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,360 @@
+
+  /**
+   * Passwords used for this user account.
+   * You can create up to two passwords for each user.
```
Added a validation.

go-to-k commented on src/aws-elasticache/README.md:
````diff
@@ -0,0 +1,345 @@
+You can import an existing user and user group by using import methods:
+
+```ts
+const importedIamUser = IamUser.fromUserAttributes(this, 'ImportedIamUser', 'my-iam-user-id');
````
If we should fix this README: 

```suggestion
const importedIamUser = IamUser.fromUserId(this, 'ImportedIamUser', 'my-iam-user-id');
```

go-to-k commented on src/aws-elasticache/README.md:
```diff
@@ -0,0 +1,345 @@
+
+This module has constructs for [Amazon ElastiCache](https://docs.aws.amazon.com/AmazonElastiCache/latest/dg/WhatIs.html).
+
+* The `User` and `UserGroup` construct facilitates the creation and management of user for the cache.
```
```suggestion
* The `User` and `UserGroup` constructs facilitate the creation and management of user for the cache.
```

go-to-k commented on src/aws-elasticache/README.md:
```diff
@@ -0,0 +1,345 @@
+
+For more information, see [Applying RBAC to a Cache for ElastiCache with Valkey or Redis OSS](https://docs.aws.amazon.com/AmazonElastiCache/latest/dg/Clusters.RBAC.html#rbac-using).
+
+If you want to create new default user, `userName` must be `default` and `userId` must not be `default` by using `NoPasswordRequiredUser` or `PasswordUser`:
```
Will CFn fail in the case of `userId` with `default` in `NoPasswordRequiredUser` or `PasswordUser`? Should we validate it?

mazyu36 replied:
Yes, the `default` user ID exists by default and can't be deleted.
<img width="1212" alt="image" src="https://github.com/user-attachments/assets/fed00978-8638-4a49-8211-72d79238216a" />

I've already added a validation to check whether the `userId` is `default` or not.

https://github.com/open-constructs/aws-cdk-library/blob/acfbe387ae58b6642b406f606a2c723f0eda45e5/src/aws-elasticache/user.ts#L159

go-to-k replied:
IAM missed it, that's OK!

Please only correct the sentence pointed out above.

mazyu36 replied:
I've corrected.

go-to-k commented on src/aws-elasticache/README.md:
````diff
@@ -0,0 +1,345 @@
+});
+```
+
+> NOTE: You can't create a new default user using `IamUser` because an IAM-enabled user's username and user ID cannot be different.
````
Should we validate it?

mazyu36 replied:
Since the validation throws an error when the `userId` is `default`, I think this requirement is already satisfied.

https://github.com/open-constructs/aws-cdk-library/blob/acfbe387ae58b6642b406f606a2c723f0eda45e5/src/aws-elasticache/user.ts#L159

mazyu36 commented on src/aws-elasticache/README.md:
```diff
@@ -0,0 +1,345 @@
+
+This user is intended for compatibility with the default behavior of previous Redis OSS versions and has an access string that permits it to call all commands and access all keys.
+
+To add proper access control to a cache, replace this default user with a new one that isn't enabled or uses a strong password.
```
I think this refers to disabling the default user by setting `off -@all` in AccessString. 
This was quoted from [AWS documentation](https://docs.aws.amazon.com/AmazonElastiCache/latest/dg/Clusters.RBAC.html#rbac-using).

<img width="1066" alt="image" src="https://github.com/user-attachments/assets/62d8c548-7b83-4864-93f4-8a45bdcb950f" />
 

However, since it may be difficult to understand, I'll consider a revision.

go-to-k commented on src/aws-elasticache/README.md:
```diff
@@ -0,0 +1,345 @@
+
+### Grant permissions to IAM-enabled users
+
+If you create IAM-enabled users, `“elasticache:Connect”` action must be allowed for user and cache.
```
```suggestion
If you create IAM-enabled users, `“elasticache:Connect”` action must be allowed for the users and cache.
```

go-to-k replied:
If the user is not an IAM user, do we need permissions to the cache?

If that story hasn't been mentioned anywhere yet, we might be able to add it (sorry if I missed it, I haven't seen the whole text yet).

mazyu36 replied:
I have made the corrections and added that no password users and password authentication users do not require permissions.

go-to-k commented on src/aws-elasticache/README.md:
```diff
@@ -0,0 +1,345 @@
+  userName: 'my-no-password-user-name',
+});
+
+const importedUserGroup = UserGroup.fromUserGroupId(this, 'ImportedUser', 'my-user-group-id');
```
```suggestion
const importedUserGroup = UserGroup.fromUserGroupId(this, 'ImportedUserGroup', 'my-user-group-id');
```

go-to-k commented on src/aws-elasticache/README.md:
```diff
@@ -0,0 +1,345 @@
+
+To control who can access the serverless cache by the security groups, use the `.connections` attribute.
+
+serverless cache have a default port `6379`.
```
```suggestion
The serverless cache have a default port `6379`.
```

go-to-k commented on src/aws-elasticache/README.md:
```diff
@@ -0,0 +1,345 @@
+
+For more information, see [Scheduling automatic backups](https://docs.aws.amazon.com/AmazonElastiCache/latest/dg/backups-automatic.html).
+
+To enable automatic backups, set the `dailySnapshotTime` property to snapshot created time:
```
How about adding the description of the snapshotRetentionLimit?

mazyu36 replied:
I've added a description about snapshotRetentionLimit.

mazyu36 commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,457 @@
+ * );
+ */
+export class PasswordUser extends BaseUser implements IPasswordUser {
+  public static fromUserAttributes(scope: Construct, id: string, attrs: PasswordUserAttributes): IPasswordUser {
```
I added a JSDoc.

mazyu36 commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,457 @@
+ * );
+ */
+export class NoPasswordRequiredUser extends BaseUser implements INoPasswordRequiredUser {
+  public static fromUserAttributes(
```
I added a JSDoc.

mazyu36 commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,457 @@
+ */
+export class PasswordUser extends BaseUser implements IPasswordUser {
+  public static fromUserAttributes(scope: Construct, id: string, attrs: PasswordUserAttributes): IPasswordUser {
+    return fromUserAttributes(scope, id, attrs);
```
Since the likelihood of additional properties being added is low, I have removed the UserAttributes and the fromUserAttributes function.

go-to-k replied:
It seemed to have not yet disappeared.
https://github.com/open-constructs/aws-cdk-library/pull/57/files#r1915975368

mazyu36 commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,457 @@
+          service: 'elasticache',
+          resource: 'user',
+          resourceName: this.userId,
+          arnFormat: ArnFormat.COLON_RESOURCE_NAME,
```
Apologies for the confusion. 

Upon rechecking, I found that all ElastiCache resources use colon-separated ARNs, so this setting was required for all resources. 

As a result, I have also added it to other places where ARNs are generated, such as in the from methods and grant methods.

https://docs.aws.amazon.com/service-authorization/latest/reference/list_amazonelasticache.html#amazonelasticache-resources-for-iam-policies

go-to-k commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,468 @@
+  readonly userName: string;
+}
+
+/**
+ * Attributes for importing a User
+ */
+export interface UserAttributes {
+  /**
+   * The ID of the user.
+   */
+  readonly userId: string;
+  /**
+   * The name of the user.
+   */
+  readonly userName: string;
+}
+
```
```suggestion
```

go-to-k commented on src/aws-elasticache/user.ts:
```diff
@@ -0,0 +1,468 @@
+import { Engine } from './util';
+
+/**
+ * A User
```
```suggestion
 * Interface for a User
```

go-to-k commented on src/aws-elasticache/user-group.ts:
```diff
@@ -0,0 +1,209 @@
+import { Engine } from './util';
+
+/**
+ * A User Group
```
Let's keep it in line with IIamUser and others, just in case.

```suggestion
 * Interface for a User Group
```

go-to-k commented on src/aws-elasticache/serverless-cache.ts:
```diff
@@ -0,0 +1,531 @@
+import { Engine } from './util';
+
+/**
+ * An ElastiCache Serverless Cache
```
Let's keep it in line with IIamUser and others, just in case.

```suggestion
 * Interface for an ElastiCache Serverless Cache
```

go-to-k commented on test/aws-elasticache/integ.elasticache-serverless-cache.ts:
```diff
@@ -0,0 +1,92 @@
+new IntegTest(app, 'ElastiCacheServerlessCacheTest', {
+  testCases: [testCase],
+  enableLookups: true,
+  stackUpdateWorkflow: false,
```
Why is `stackUpdateWorkflow` false?

Is there also a reason for explicitly specifying `enableLookups` (with true)?

mazyu36 replied:
It was something that remained when copied from another file, and I deleted it since it was unnecessary.

go-to-k commented on test/aws-elasticache/integ.elasticache-serverless-cache.ts:
```diff
@@ -0,0 +1,92 @@
+      userName: 'default',
+    });
+
+    const iamUser = new ocf.aws_elasticache.IamUser(this, 'IamUser', {
+      accessString: 'on ~* +@all',
+    });
```
Could you please do the following?

- Import a user
- Grant the imported user
- Add the imported user to a user group

Perhaps an integ test file for the user and user group could be prepared.

go-to-k replied:
> Or, perhaps an integ test file for the User could be prepared.

If we need to add the user group with the user to a serverless cache to make it a meaningful test, we can put it in this file.

mazyu36 replied:
I've updated an integ test.

However, another issue has occurred and I am investigating it.
https://github.com/open-constructs/aws-cdk-library/pull/57/commits/ddd90fac3cbe55dbc22f65ad92f9dcd1875f91bd#r1916135486

go-to-k commented on src/aws-elasticache/README.md:
```diff
@@ -0,0 +1,348 @@
+
+### Grant permissions to IAM-enabled users
+
+If you create IAM-enabled users, `“elasticache:Connect”` action must be allowed for the users and cache.
```
```suggestion
If you create IAM-enabled users, `"elasticache:Connect"` action must be allowed for the users and cache.
```

go-to-k commented on src/aws-elasticache/README.md:
```diff
@@ -0,0 +1,348 @@
+declare const serverlessCache: ServerlessCache;
+declare const role: iam.Role;
+
+// grant “elasticache:Connect” action permissions to role
```
```suggestion
// grant "elasticache:Connect" action permissions to role
```

go-to-k reviewed (commented):
More general would be better.

go-to-k commented on src/aws-elasticache/README.md:
````diff
@@ -0,0 +1,348 @@
+});
+```
+
+### Connecting to serverless cache
````
How about talking about the grant method here?

go-to-k replied:
Sorry, I missed that the IAM auth for the cache required with the IAM user.

go-to-k commented on test/aws-elasticache/integ.elasticache-serverless-cache.ts.snapshot/ElastiCacheServerlessCacheStack.template.json:
```diff
@@ -20,8 +20,8 @@
     "Engine": "redis",
-    "UserId": "elasticacheservehestack-iamuser-1980b570",
-    "UserName": "elasticacheservehestack-iamuser-1980b570"
+    "UserId": "elasticacheserverlesscachestack",
```
If the scope (the first arg in the constructor) is stack, `makeUniqueResourceName` in `Names.uniqueResourceName` returns the stack name.

In other words, if the physicalName is passed into an userId, the userId will be the stack name.

https://github.com/go-to-k/aws-cdk/blob/946b7488e680051ccdecb6a7f932d98ab6b6ce6f/packages/aws-cdk-lib/core/lib/private/unique-resource-name.ts#L66-L75

Can we use `PhysicalName.GENERATE_IF_NEEDED` for the physicalName?

```ts
  protected constructor(scope: Construct, id: string, props: BaseUserProps = {}) {
    super(scope, id, {
      physicalName:
        props.userId ?? PhysicalName.GENERATE_IF_NEEDED,
    });
```

go-to-k replied:
Sorry, that didn't work in that case. I'll look at it again later when I have more time.

mazyu36 replied:
Thank you. I've been looking into the source code for uniqueResourceName on my end as well, but haven't been able to pinpoint the cause yet.

When I went back and re-ran the integration tests with older versions of the source code, it appears the behavior changed when the constructor was consolidated into the base class in the following commit.

https://github.com/open-constructs/aws-cdk-library/commit/f95831e06e036bfabe0cc7b26c52f7a22d23b4ae

When running the integration tests using the code from the earlier commit below, the userId included the construct id as well.

https://github.com/open-constructs/aws-cdk-library/commit/dda185a753051162fde18a73d8cb3e72e45e5b1f#diff-c93c1ee4d52b64660eb4aabb473c9a793300c62459e005565025587f78776d11

```bash
Resources
[~] AWS::ElastiCache::User IamUser83DA77ED replace
 ├─ [~] UserId (requires replacement)
 │   ├─ [-] elasticacheserverlesscachestack
 │   └─ [+] elasticacheservehestack-iamuser-1980b570
 └─ [~] UserName (requires replacement)
     ├─ [-] elasticacheserverlesscachestack
     └─ [+] elasticacheservehestack-iamuser-1980b570
```

mazyu36 replied:
Ahhh, I made an implementation mistake during refactoring....
I've fixed it. Sorry for such a basic mistake.

go-to-k replied:
Thank you!

go-to-k commented:
Did we get all the threads resolved? (let me know if I missed any)

I think I've seen everything else, nothing appears to be wrong (which is so great).
I'll merge this PR if you are OK, but I'd like you to give me a final check just to be sure.

mazyu36 commented:
Thank you. 
I have addressed all the comments, but I will double-check everything tomorrow and get back to you.

mazyu36 commented:
@go-to-k 
I fixed several typos in the README.
I also improved the explanation and implementation samples for the original default user that is automatically created in AWS. 
Additionally, I added an integration test for the original default user.

Please let me know if this looks good to you.

go-to-k commented:
Looks good. May we merge this PR?

mazyu36 commented:
Thank you. I would appreciate it if you could proceed with the merge.
